stdlib/public/Concurrency/AsyncStreamBackpressure.swift

```diff
@@ -0,0 +1,1961 @@
+
+                case .resumeConsumerWithElementAndProducers(let continuation, let element, let producerContinuations):
+                    continuation.resume(returning: element)
+                    for producerContinuation in producerContinuations {
+                        producerContinuation(Result<Void, Error>.success(()))
+                    }
+
+                case .resumeConsumerWithFailureAndCallOnTermination(let continuation, let failure, let onTermination):
+                    switch failure {
+                    case .some(let error):
+                        continuation.resume(throwing: error)
+
+                    case .none:
+                        continuation.resume(returning: nil)
+                    }
+                    onTermination?()
+
+                case .resumeConsumerWithNil(let continuation):
+                    continuation.resume(returning: nil)
+
+                case .none:
+                    break
+                }
+            }
+        } onCancel: {
+            let action = self._stateMachine.withCriticalRegion {
+                $0.cancelNext()
+            }
+
+            switch action {
+            case .resumeConsumerWithCancellationErrorAndCallOnTermination(let continuation, let onTermination):
+                continuation.resume(throwing: CancellationError())
+                onTermination?()
+
+            case .failProducersAndCallOnTermination(let producerContinuations, let onTermination):
+                for producerContinuation in producerContinuations {
+                    producerContinuation(.failure(AsyncStreamAlreadyFinishedError()))
+                }
+                onTermination?()
+
+            case .none:
+                break
+            }
+        }
+    }
+}
+
+/// The state machine of the backpressured async stream.
+@available(SwiftStdlib 5.1, *)
+struct _StateMachine<Element, Failure: Error> {
+    enum _State {
+        struct Initial {
+            /// The backpressure strategy.
+            var backPressureStrategy: _InternalBackPressureStrategy
+            /// Indicates if the iterator was initialized.
+            var iteratorInitialized: Bool
+            /// The onTermination callback.
+            var onTermination: (@Sendable () -> Void)?
+        }
+
+        struct Streaming {
+            /// The backpressure strategy.
+            var backPressureStrategy: _InternalBackPressureStrategy
+            /// Indicates if the iterator was initialized.
+            var iteratorInitialized: Bool
+            /// The onTermination callback.
+            var onTermination: (@Sendable () -> Void)?
+            /// The buffer of elements.
+            var buffer: _Deque<Element>
+            /// The optional consumer continuation.
+            var consumerContinuation: CheckedContinuation<Element?, Error>?
+            /// The producer continuations.
+            var producerContinuations: _Deque<(UInt, (Result<Void, Error>) -> Void)>
```

Comment:
I wonder if the Uint token could instead be an index into the deque? That way there is no need for lookup?
The one issue with that is we would have to ensure invariance for the stored deque while that index is held.

Reply:
We can't use an index into the Deque here due to two things:
1. We have to generate the token before we get an actual continuation which would mean we would have to reserve a slot 
2. Between the generation of the token and the call to enqueue/cancel the producer anything can interleave

Due to the above two constraints I think it is best if we generated a monotonically increasing token and we just have to search for the correct producer. In most cases, we are only expecting a single producer or maybe a few anyhow.